cabal-install/bootstrap.py

```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+bootstrap.py - bootstrapping utility for cabal-install.
+
+This utility is intended for use (only) in building cabal-install
+on a new platform.
+
+Usage:
+
+  1. On a system with functional cabal-install, use `cabal v2-configure` to
+     compute an install plan for the same GHC version that you will be using to
+     bootstrap.
+
+  2. Build a dependency description file (bootstrap-deps.json) from this
+     install plan by running `bootstrap.py --extract-plan`.
+
+  3. Copy bootstrap-deps.json to the bootstrapping environment.
+
+  4. On the system you are bootstrapping, run `bootstrap.py -d
+     bootstrap-deps.json`
+"""
+
+from enum import Enum
+import hashlib
+import json
+from pathlib import Path
+import shutil
+import subprocess
+from textwrap import dedent
+from typing import Set, Optional, Dict, List, Tuple, \
+                   NewType, BinaryIO, NamedTuple
+
+ghc_path = Path('ghc')
```

Comment:
One of problems with old bootstrap was that we were somewhat sensitive to the `ghc` version used. We should use `ghc --version` info.

Or better we should parse `ghc --info` and find `ghc-pkg` using `LibDir`, as perfectly boostrapping ghc should be configurable (i.e. we wouldn't need to rely on PATH).

Reply:
> Or better we should parse `ghc --info` and find `ghc-pkg` using `LibDir`, as perfectly boostrapping ghc should be configurable (i.e. we wouldn't need to rely on PATH).

This suggestion seems very fragile to me. The `ghc-pkg` wrapper in `bin/` arranges to pass `--global-package-db` to the `ghc-pkg` executable in `LibDir`. We would need to duplicate this logic if we were to use the `LibDir` executable directly.

Moreover, on some platforms (e.g. NixOS) the executable in `PATH` is actually a distribution-provided wrapper which sets up other state. IMHO we shouldn't start poking around in `LibDir`. If anything we should just look for a `ghc-pkg` in the same directory as we found the `ghc` executable.

Reply:
It actually turns out the that the `--global-package-db` flag that I noted above is added by the NixOS wrapper. Nevertheless, the point stands.

Reply:
I'm sad to learn about this.

The simple heuristics don't work:

```
% $(dirname $(which ghc-8.10.1))/ghc-pkg --version
GHC package manager version 8.8.3
```

we have better heuristics for these, but they don't work if suffix is arbitrary. I'd rather not to use heuristics, and for special (and I consider NixOS special) situations users would rather required to be explict.

NixOS probably uses wrapper because `--global-package-db` because they fake global installs, not actually do that. That doesn't affect bootstrap.sh because it doesn't use "extras".